subsys/testsuite/busy_sim/busy_sim.c

```diff
@@ -0,0 +1,162 @@
+
+struct busy_sim {
+	uint32_t idle_avg;
+	uint32_t active_avg;
+	uint16_t idle_delta;
+	uint16_t active_delta;
+	uint32_t us_tick;
+	const struct device *entropy;
+	const struct device *counter;
+	const struct device *gpio_port;
+	struct counter_alarm_cfg alarm_cfg;
+	uint8_t pin;
+};
+
+static struct busy_sim sim = {
+	.counter = DEVICE_DT_GET(DT_PARENT(DT_NODELABEL(busy_sim))),
+#if DT_NODE_HAS_PROP(DT_NODELABEL(busy_sim), gpios)
+	.gpio_port = DEVICE_DT_GET(DT_GPIO_CTLR(DT_NODELABEL(busy_sim), gpios)),
+	.pin = DT_GPIO_PIN(DT_NODELABEL(busy_sim), gpios)
+#endif
+};
+
+#define BUFFER_SIZE 32
+RING_BUF_DECLARE(rbuf, BUFFER_SIZE);
+static void rng_pool_work_handler(struct k_work *work)
+{
+	uint8_t *data;
+	uint32_t len;
+
+	len = ring_buf_put_claim(&rbuf, &data, BUFFER_SIZE - 1);
+	if (len) {
+		int err = entropy_get_entropy(sim.entropy, data, len);
+
+		if (err == 0) {
+			ring_buf_put_finish(&rbuf, len);
+			return;
+		}
+	}
+
+	k_work_submit(work);
+}
+
+static K_WORK_DEFINE(rng_pool_work, rng_pool_work_handler);
+
+static uint32_t get_timeout(bool idle)
+{
+	uint32_t avg = idle ? sim.idle_avg : sim.active_avg;
+	uint32_t delta = idle ? sim.idle_delta : sim.active_delta;
+	uint16_t rand_val;
+	uint32_t len;
+
+	len = ring_buf_get(&rbuf, (uint8_t *)&rand_val, sizeof(rand_val));
+	if (len < sizeof(rand_val)) {
+		k_work_submit(&rng_pool_work);
+		rand_val = 0;
+	}
+
+	avg *= sim.us_tick;
+	delta *= sim.us_tick;
+
+	return avg - delta + 2 * (rand_val % delta);
+}
+
+static void counter_alarm_callback(const struct device *dev,
+				   uint8_t chan_id, uint32_t ticks,
+				   void *user_data)
+{
+	int err;
+
+	sim.alarm_cfg.ticks = get_timeout(true);
+
+	if (sim.gpio_port) {
+		err = gpio_pin_set(sim.gpio_port, sim.pin, 1);
+		__ASSERT_NO_MSG(err >= 0);
+	}
+
+	/* Busy loop */
+	k_busy_wait(get_timeout(false) / sim.us_tick);
+
+	if (sim.gpio_port) {
+		err = gpio_pin_set(sim.gpio_port, sim.pin, 0);
+		__ASSERT_NO_MSG(err >= 0);
+	}
+
+	err = counter_set_channel_alarm(sim.counter, 0, &sim.alarm_cfg);
+	__ASSERT_NO_MSG(err == 0);
+
+}
+
+void busy_sim_start(uint32_t active_avg, uint32_t active_delta,
+		    uint32_t idle_avg, uint32_t idle_delta)
+{
+	int err;
+
+	sim.active_avg = active_avg;
+	sim.active_delta = active_delta;
+	sim.idle_avg = idle_avg;
+	sim.idle_delta = idle_delta;
+
+	err = k_work_submit(&rng_pool_work);
+	__ASSERT_NO_MSG(err >= 0);
+
+	sim.alarm_cfg.ticks = counter_us_to_ticks(sim.counter, 100);
+	err = counter_set_channel_alarm(sim.counter, 0, &sim.alarm_cfg);
+	__ASSERT_NO_MSG(err == 0);
+
+	err = counter_start(sim.counter);
+	__ASSERT_NO_MSG(err == 0);
+}
+
+void busy_sim_stop(void)
+{
+	int err;
+
+	k_work_cancel(&rng_pool_work);
+	err = counter_stop(sim.counter);
+	__ASSERT_NO_MSG(err == 0);
+}
+
+static int busy_sim_init(const struct device *unused)
+{
+	ARG_UNUSED(unused);
+	uint32_t freq;
+
+	if (sim.gpio_port) {
+		gpio_pin_configure(sim.gpio_port, sim.pin,
+				   GPIO_OUTPUT | GPIO_OUTPUT_INIT_LOW);
```

Comment:
This way the flags specified in DT for this pin are ignored (what may be confusing). It would be better to use `GPIO_DT_SPEC_GET()` and `gpio_pin_configure_dt()` for handling the pin. See for instance the button sample: https://github.com/zephyrproject-rtos/zephyr/blob/bf3c6e7ba415dd85f1b68eb69ea2779b234c686f/samples/basic/button/src/main.c#L32-L33 https://github.com/zephyrproject-rtos/zephyr/blob/bf3c6e7ba415dd85f1b68eb69ea2779b234c686f/samples/basic/button/src/main.c#L75-L76.